# Linear Regression

#### Week 3 - Day 2

In this notebook, we will cover:
1. Train a **LinearRegression** model on a provided regression dataset (I used California housing).
2. Report the model's **coefficients** and identify which **feature** has the strongest effect.
3. Evaluate the model with **MAE**, **RMSE**, and **R²** on the test set.
4. Compare the RMSE against a **baseline** that predicts the mean for every row, and state whether the model adds value.
5. Document the interpretation of results in Markdown.
   > Note: Words written in bold represents key topics that were studied and covered during the day.

## Step 0 - Setup data & Load the dataset:

In [8]:
import pandas as pd
import numpy as np
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

data = fetch_california_housing(as_frame=True)
df = data.frame
df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


## Step 1.1 - Separate the features & the Target:

In [2]:
X = df.drop("MedHouseVal", axis=1)   # means chose all columns without "MedHouseVal" -> The features.
y = df["MedHouseVal"]                # -> The target.

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (20640, 8)
y shape: (20640,)


## Step 1.2 - train/test Split & Evaluate the model:

In [5]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42  # test_size=0.2: means holds out 20% for testing.  /  random_state=42: means fixes the split so results are reproducible.
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)


model = LinearRegression()
model.fit(X_train, y_train)

predictions = model.predict(X_test)

X_train: (16512, 8)
X_test: (4128, 8)
y_train: (16512,)
y_test: (4128,)


## Step 2 - The coefficients:

In [6]:
coef_df = pd.DataFrame({
    "Feature": X.columns,
    "Coefficient": model.coef_
}).sort_values(by="Coefficient", key=abs, ascending=False)

print(coef_df)
print("\nIntercept:", model.intercept_)

      Feature  Coefficient
3   AveBedrms     0.783145
0      MedInc     0.448675
7   Longitude    -0.433708
6    Latitude    -0.419792
2    AveRooms    -0.123323
1    HouseAge     0.009724
5    AveOccup    -0.003526
4  Population    -0.000002

Intercept: -37.02327770606416


## Step 3 - Evaluate the model with MAE, RMSE and R²:

In [9]:
mae = mean_absolute_error(y_test, predictions)
rmse = np.sqrt(mean_squared_error(y_test, predictions))
r2 = r2_score(y_test, predictions)

print(f"MAE: {mae:.4f}")
print(f"RMSE: {rmse:.4f}")
print(f"R²: {r2:.4f}")

MAE: 0.5332
RMSE: 0.7456
R²: 0.5758


## Step 4 - Compare the RMSE against a baseline:

In [10]:
baseline_pred = np.full_like(y_test, y_train.mean())
baseline_rmse = np.sqrt(mean_squared_error(y_test, baseline_pred))

print(f"Baseline RMSE (predicting mean): {baseline_rmse:.4f}")
print(f"Model RMSE: {rmse:.4f}")
print(f"Improvement: {baseline_rmse - rmse:.4f}")

Baseline RMSE (predicting mean): 1.1449
Model RMSE: 0.7456
Improvement: 0.3993


## Step 5 - Interpretation of results:

- The strongest feature affecting house price is **MedInc** (median income), with the highest coefficient — meaning higher income areas strongly predict higher house values.
- **The model's RMSE (0.7456)** is significantly **lower than** the **baseline RMSE (1.1449)** (predicting the mean for every row), showing the model captured real patterns instead of just guessing the average.
- **R² of 0.5758** means **the model explains about 60% of the variance in house prices** — a reasonable result for a simple linear model on this dataset.

> **Note: Every cell of above is explaied by markdown cell or by comments.**